File: src/main/java/com/paragon/client/ui/configuration/old/OldPanelGUI.java
package com.paragon.client.ui.configuration.old;

import com.paragon.Paragon;
import com.paragon.api.module.Category;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.font.FontUtil;
import com.paragon.client.systems.module.impl.client.ClickGUI;
import com.paragon.client.systems.module.impl.client.ClientFont;
import com.paragon.client.systems.module.impl.client.Colours;
import com.paragon.client.ui.configuration.GuiImplementation;
import com.paragon.client.ui.configuration.old.impl.Panel;
import org.lwjgl.input.Mouse;

import java.util.ArrayList;
import java.util.Objects;

/**
 * @author Wolfsurge
 */
public final class OldPanelGUI extends GuiImplementation {

    public static OldPanelGUI INSTANCE = new OldPanelGUI();

    // The tooltip being rendered
    public static String tooltip = "";
    // List of panels
    private final ArrayList<Panel> panels = new ArrayList<>();

    private OldPanelGUI() {
        // X position of panel
        float x = (RenderUtil.getScreenWidth() / 2) - ((Category.values().length * 100) / 2f);

        // Add a panel for every category
        for (Category category : Category.values()) {
            // Add panel
            panels.add(new Panel(x, 30, 95, 16, category));

            // Increase X
            x += 100;
        }
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, int mouseDelta) {
        // Reset tooltip
        tooltip = "";

        // Make the background darker
        if (ClickGUI.getDarkenBackground().getValue()) {
            drawDefaultBackground();
        }

        for (Panel panel : panels) {
            panel.setY(panel.getY() + (mouseDelta / 100f) * ClickGUI.getScrollSpeed().getValue());
        }

        // Render panels
        panels.forEach(panel -> {
            panel.renderPanel(mouseX, mouseY);
        });

        if (ClickGUI.getTooltips().getValue() && !Objects.equals(tooltip, "")) {
            RenderUtil.drawRect(mouseX + 7, mouseY - 5, FontUtil.getStringWidth(tooltip) + 4, FontUtil.getHeight() + 2, 0x90000000);
            RenderUtil.drawBorder(mouseX + 7, mouseY - 5, FontUtil.getStringWidth(tooltip) + 4, FontUtil.getHeight() + 2, 0.5f, Colours.mainColour.getValue().getRGB());
            FontUtil.drawStringWithShadow(tooltip, mouseX + 9, mouseY - (ClientFont.INSTANCE.isEnabled() ? 2 : 4), -1);
        }
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        // Clicks
        panels.forEach(panel -> {
            panel.mouseClicked(mouseX, mouseY, mouseButton);
        });
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, int state) {
        // Click releases
        panels.forEach(panel -> {
            panel.mouseReleased(mouseX, mouseY, state);
        });
    }

    @Override
    public void keyTyped(char typedChar, int keyCode) {
        // Keys typed
        panels.forEach(panel -> {
            panel.keyTyped(typedChar, keyCode);
        });
    }

    @Override
    public void onGuiClosed() {
        Paragon.INSTANCE.getStorageManager().saveModules("current");
    }

    public void scrollPanels() {
        int dWheel = Mouse.getDWheel();

        for (Panel panel : panels) {
            panel.setY(panel.getY() + (dWheel / 100f) * ClickGUI.getScrollSpeed().getValue());
        }
    }

    @Override
    public boolean doesGuiPauseGame() {
        // Pause the game if pause is enabled in the GUI settings
        return ClickGUI.getPause().getValue();
    }

    /**
     * Gets the panels
     *
     * @return The panels
     */
    public ArrayList<Panel> getPanels() {
        return panels;
    }

    public static boolean isInside(float x, float y, float x2, float y2, int checkX, int checkY) {
        return checkX >= x && checkX <= x2 && checkY > y && checkY <= y2;
    }

}

File: src/main/kotlin/com/paragon/client/ui/configuration/old/OldPanelGUI.kt
package com.paragon.client.ui.configuration.old

import com.paragon.Paragon
import com.paragon.api.module.Category
import com.paragon.api.util.render.RenderUtil.drawBorder
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.api.util.render.RenderUtil.screenWidth
import com.paragon.api.util.render.font.FontUtil.drawStringWithShadow
import com.paragon.api.util.render.font.FontUtil.getHeight
import com.paragon.api.util.render.font.FontUtil.getStringWidth
import com.paragon.client.systems.module.impl.client.ClickGUI.darkenBackground
import com.paragon.client.systems.module.impl.client.ClickGUI.pause
import com.paragon.client.systems.module.impl.client.ClickGUI.scrollSpeed
import com.paragon.client.systems.module.impl.client.ClickGUI.tooltips
import com.paragon.client.systems.module.impl.client.ClientFont
import com.paragon.client.systems.module.impl.client.Colours
import com.paragon.client.ui.configuration.GuiImplementation
import com.paragon.client.ui.configuration.old.impl.Panel
import org.lwjgl.input.Mouse
import java.util.function.Consumer

/**
 * @author Wolfsurge
 */
class OldPanelGUI private constructor() : GuiImplementation() {

    // List of panels
    val panels = ArrayList<Panel>()

    init {
        // X position of panel
        var x = screenWidth / 2 - Category.values().size * 100 / 2f

        // Add a panel for every category
        for (category in Category.values()) {
            // Add panel
            panels.add(Panel(x, 30f, 95f, 16f, category))

            // Increase X
            x += 100f
        }
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, mouseDelta: Int) {
        // Reset tooltip
        tooltip = ""

        for (panel in panels) {
            panel.y = panel.y + mouseDelta / 100f * scrollSpeed.value
        }

        // Render panels
        panels.forEach(Consumer { panel: Panel -> panel.renderPanel(mouseX, mouseY) })
        if (tooltips.value && tooltip != "") {
            drawRect((mouseX + 7).toFloat(), (mouseY - 5).toFloat(), getStringWidth(tooltip) + 4, getHeight() + 2, -0x70000000)
            drawBorder((mouseX + 7).toFloat(), (mouseY - 5).toFloat(), getStringWidth(tooltip) + 4, getHeight() + 2, 0.5f, Colours.mainColour.value.rgb)
            drawStringWithShadow(tooltip, (mouseX + 9).toFloat(), (mouseY - if (ClientFont.isEnabled) 2 else 4).toFloat(), -1)
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        // Clicks
        panels.forEach(Consumer { panel: Panel -> panel.mouseClicked(mouseX, mouseY, mouseButton) })
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, state: Int) {
        // Click releases
        panels.forEach(Consumer { panel: Panel -> panel.mouseReleased(mouseX, mouseY, state) })
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        // Keys typed
        panels.forEach(Consumer { panel: Panel -> panel.keyTyped(typedChar, keyCode) })
    }

    override fun onGuiClosed() {
        Paragon.INSTANCE.storageManager.saveModules("current")
    }

    fun scrollPanels() {
        val dWheel = Mouse.getDWheel()
        for (panel in panels) {
            panel.y = panel.y + dWheel / 100f * scrollSpeed.value
        }
    }

    override fun doesGuiPauseGame(): Boolean {
        // Pause the game if pause is enabled in the GUI settings
        return pause.value
    }

    companion object {
        var INSTANCE = OldPanelGUI()

        // The tooltip being rendered
        @JvmField
        var tooltip = ""
        @JvmStatic
        fun isInside(x: Float, y: Float, x2: Float, y2: Float, checkX: Int, checkY: Int): Boolean {
            return checkX >= x && checkX <= x2 && checkY > y && checkY <= y2
        }
    }
}